### PR TITLE
Enhanced Python 3 Support

### PR DESCRIPTION
**Pull request by gthank**
I just set up a new instance of ReadTheDocs under Python 3.5 on FreeBSD. These are the changes I needed to make to complete the [installation instructions](https://read-the-docs.readthedocs.io/en/latest/install.html). There are probably additional changes required to achieve comprehensive Python 3 support, but I hope this is seen as a good start.

### CODE REVIEW

**Comment by agjohnson** [comment]
Thanks for giving this a go! The larger issue is all of the dependencies that we have that are python 2 only and perhaps unmaintained. This is where I stopped last time I was working on python 3 support.

This PR is going to be prone to breakage while it is still a WIP, as we can't easily do this change incrementally.

**Comment by gthank** [comment]
Dependencies are obviously a potential source of headaches. I can report that as of right now, the changes I submitted are enough to get RTD installed and running under Gunicorn on FreeBSD 11 w/ Python 3.5. I've so far managed to sign up a new user (though it won't stay logged in; I suspect I've missed some config overrides) and am not seeing any actual exceptions.

Since I'm only self-hosting, that's probably not enough to test all the integrations, etc. and flush out the bugs, but the changes I've submitted so far should be fairly safe for doing single codebase. It's consisted mostly of using `six` to handle the move of `urlparse`, a bunch of `__future__` imports to get the `print()` function (and subsequent usage of it instead of the `print` statement), and a couple of tweaks to the requirements file (one to drop distutils2, since it seems to be obsolete, and a version bump to `slumber` to get on a version that would install under 3.5.

I'd be happy  to work through whatever process is appropriate for getting these changes more fully tested with the goal of getting the PR accepted, as well as to submit more enhancements to fix issues I find while running ReadTheDocs locally.

--Hank

**Comment by agjohnson** [comment]
Great! I'll check in next week, let us know how testing is going on this. As I recall, there were some old django packages that weren't looking like they would get python3 support -- slumber was the main package lacking python3 support though. Perhaps that has changed. The packages that were going to be a problem also weren't heavily used, so it might introduce some edge case bugs.

It would help to have a core dev pull this down to run tests and QA once it's about ready.

**Comment by gthank** [comment]
On the subject of dependencies, I found a tool—[checkmyreqs](https://pypi.python.org/pypi/checkmyreqs/0.3.1)—to automatically check the entries in a `requirements.txt` to see if they are tagged with a given Python version tag on PyPI. I had to fork it to use HTTPS when talking to PyPI, but after I did that, a check against the dependencies is fairly encouraging:

```
$ checkmyreqs -p 3.4 -f requirements/pip.txt
Checking for compatibility with Python 3.4
requirements/pip.txt
*****
pip
compatible
-----
appdirs
compatible
-----
virtualenv
compatible
-----
docutils
compatible
-----
Sphinx
compatible
-----
sphinx_rtd_theme
compatible
-----
Pygments
compatible
-----
mkdocs
compatible
-----
readthedocs-build
not specified
-----
django
not listed on pypi.python.org
-----
django-tastypie
compatible
-----
django-haystack
compatible
-----
celery-haystack
not compatible - update to v0.10 for support
-----
django-guardian
compatible
-----
django-extensions
compatible
-----
djangorestframework
compatible
-----
django-vanilla-views
compatible
-----
jsonfield
compatible
-----
pytest-django
compatible
-----
requests
compatible
-----
slumber
not specified
-----
lxml
compatible
-----
django-countries
compatible
-----
redis
compatible
-----
celery
compatible
-----
django-celery
compatible
-----
dnspython
not specified - update to v1.15.0 for explicit support
-----
httplib2
compatible
-----
elasticsearch
compatible
-----
pyelasticsearch
compatible
-----
pyquery
compatible
-----
django-gravatar2
compatible
-----
doc2dash
compatible
-----
pytz
compatible
-----
beautifulsoup4
compatible
-----
Unipath
not specified - update to v1.1 for explicit support
-----
django-kombu
not specified
-----
django-secure
not compatible
-----
mimeparse
not specified
-----
mock
compatible
-----
stripe
compatible
-----
django-copyright
not specified
-----
django-formtools
compatible
-----
django-dynamic-fixture
compatible
-----
docker-py
compatible
-----
django-textclassifier
compatible
-----
django-annoying
compatible
-----
django-messages-extends
not specified
-----
django-filter
compatible
-----
djangorestframework-jsonp
compatible
-----
sphinxcontrib-httpdomain
not specified
-----
commonmark
not listed on pypi.python.org
-----
recommonmark
not specified
-----
packaging
compatible
-----
nilsimsa
not specified
-----
```

**Comment by ericholscher** [comment]
I believe we should be able to slowly upgrade the codebase to support Python 3 without breaking it on Python 2. Things like `print()` and whatnot are supported in 2.7, so I think thats a good way to start moving forward.

**Comment by ericholscher** [comment]
Got this running locally:

```
========================= 26 failed, 408 passed, 1 xfailed, 993 warnings in 75.20 seconds ==========================
```

Will take a peek.

**Comment by ericholscher** [comment]
This got me down to 17 failures:

```
diff --git a/readthedocs/projects/version_handling.py b/readthedocs/projects/version_handling.py
index fcb725a..2a69c9f 100644
--- a/readthedocs/projects/version_handling.py
+++ b/readthedocs/projects/version_handling.py
@@ -113,7 +113,7 @@ def version_windows(versions, major=1, minor=1, point=1):
 def parse_version_failsafe(version_string):
     try:
         return Version(
-            text_type(unicodedata.normalize('NFKD', text_type(version_string)).encode('ascii', 'ignore'))
+            text_type(unicodedata.normalize('NFKD', text_type(version_string)))
         )
     except (UnicodeError, InvalidVersion):
         return None
```

**Comment by ericholscher** [comment]
Pushed a couple fixes, which got my tests down to ~15 failing. Definitely still need to go through and clean up our unicode fails :)

**Comment by gthank** [comment]
Closing this one because I made a different one that is way better.
